dig.go

```diff
@@ -210,14 +259,69 @@ func (c *Container) get(t reflect.Type) (reflect.Value, error) {
 
 	for _, con := range constructed {
 		ct := con.Type()
-		if ct == _errType {
-			continue
-		}
-		c.cache[ct] = con
+		c.set(ct, con)
 	}
 	return c.cache[t], nil
 }
 
+// Returns a new In parent object with all the dependency fields
+// populated from the dig container.
+func (c *Container) createInObject(t reflect.Type) (reflect.Value, error) {
+	dest := reflect.New(t).Elem()
+	for i := 0; i < t.NumField(); i++ {
+		f := t.Field(i)
+		if f.PkgPath != "" {
+			continue // skip private fields
+		}
+
+		var isOptional bool
+		if tag := f.Tag.Get(_optionalTag); tag != "" {
+			var err error
+			isOptional, err = strconv.ParseBool(tag)
+			if err != nil {
+				return dest, fmt.Errorf(
+					"invalid value %q for %q tag on field %v of %v: %v",
+					tag, _optionalTag, f.Name, t, err)
+			}
+		}
+
+		v, err := c.get(f.Type)
+		if err != nil {
+			if isOptional {
+				v = reflect.Zero(f.Type)
+			} else {
+				return dest, fmt.Errorf(
+					"could not get field %v (type %v) of %v: %v", f.Name, f.Type, t, err)
+			}
+		}
+
+		dest.Field(i).Set(v)
+	}
+	return dest, nil
+}
+
+// Set the value in the cache after a node resolution
+func (c *Container) set(t reflect.Type, v reflect.Value) {
+	if !isOutObject(t) {
+		// do not cache error types
+		if t != _errType {
+			c.cache[t] = v
+		}
+		return
+	}
+
+	// dig.Out objects are not acted upon directly, but rather their memebers are considered
+	for i := 0; i < t.NumField(); i++ {
+		var (
+			field = t.Field(i)
+			ft    = field.Type
+			fv    = v.Field(i)
+		)
+		// recurse into all embedded objects
```

Comment:
nit: these are fields, but they're not necessarily embedded. https://golang.org/ref/spec#Struct_types